nssdb/keys.go

```diff
@@ -182,14 +195,27 @@ func (db *NSSDB) AddPrivateKey(ctx context.Context, privKey *ecdsa.PrivateKey, n
 }
 
 // AddPublicKey adds a public key to the nssPublic database and returns its id.
+// The ckaID argument should come from the SubjectKeyID of the associated certificate.
+// Keys with the same ckaID will be replaced.
 // Only ecdsa keys with curve P-256 are supported.
-// TODO(areed) what if a key already has the cka id? Update it?
-func (db *NSSDB) AddPublicKey(ctx context.Context, pubKey *ecdsa.PublicKey, id []byte) (uint32, error) {
+func (db *NSSDB) AddPublicKey(ctx context.Context, pubKey *ecdsa.PublicKey, ckaID []byte) (uint32, error) {
 	if pubKey.Curve != elliptic.P256() {
 		return 0, errors.New("unsupported curve")
 	}
 
-	pubKeyObj, err := ecPubKeyToObject(pubKey, id)
+	if len(ckaID) > 0 {
+		matches, err := db.findByAttr(ctx, CKO_PUBLIC_KEY, "CKA_ID", ckaID)
+		if err != nil {
+			return 0, fmt.Errorf("find cka id conflicts: %w", err)
+		}
+		for _, id := range matches {
+			if err := db.DeleteObjectPublic(ctx, id); err != nil {
+				return 0, fmt.Errorf("delete conflicting public key %d: %w", id, err)
+			}
```

Comment:
Per before, replacing objects with the same id or label might make sense, but it's also an option to fail, and let the user delete it if it already exists. We do that in the KMS, even in `step`, where the decision is taken by the user, either typing yes or passing the `--force` flag.

I think we should consider what the cli utils that are used to manage NSS databases do by default.

Reply:
Keys are a bit different than certificates. `pk12util` replaces the certificate with the same CKA_ID but for keys it leaves the old ones in place. This package deletes and re-creates so the keys will have updated ids but the attributes should be the same.